### ISSUE TITLE
localhost mapping

### ISSUE THREAD

**Issue by biqas** [dotnet/aspire#2811]
Hello,

I have finally tried aspire, was hard to setup in my case because I am developing on an Apple Silicon (M1) and using Parallels Windows ARM for Visual Studio and Linux (Fedora) for Podman.

And The problem I have is I have setup on mac in etc/host the DNS for Fedora Linux so I can access in the Windows VM the Podman containers via DNS. On command line everything is fine, and the environment works fine.

```
MAC etc/hosts file:
10.211.66.55     fedora.podman
```

I have not found any possibility to map in Aspire `localhost` to `fedora.podman` for resource like Postgres (which is running as a container in fedora Linux as a Podman container) or any other resource.

What i try to do is something like this:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

// Secrets
var postgresPassword = builder.Configuration["postgres.password"];


// Resources
var postgres = builder.AddPostgres("postgres", password: postgresPassword)
    .WithImageTag("16.2");

var dbA = postgres.AddDatabase("db-a", "db.a");
var dbB = postgres.AddDatabase("Db-b", "db.b");


// Apps
builder.AddProject<Projects.SOME_PROJECT>("some-project")
    .WithReference(dbA)
    .WithReference(dbB);

builder.Build().Run();
```

this results in that the connection string is like this:
`Server=localhost;Port=5432;Database=db.a;User ID=XXX;Password='XXX'`

I need to override for this resource to not use `localhost` but instead `fedora.podman`.

Is this missing or not supported?

**Comment by davidfowl** [comment]
This isn't possible today. The only feature we have today around localhost mapping is when the containers need to access resources on the host, we have a setting that replaces localhost with host.docker.internal (which is configurable).

This is similar but because your container host is running remotely. We'd need another setting for host -> container host.

**Comment by davidfowl** [comment]
Tangentially related to https://github.com/dotnet/aspire/issues/2639

**Comment by biqas** [comment]
Would it make sense then to introduce a broader concept here, which can be also leveraged by other future tools.

The concept would be to put resource into groups or so called "containers" (not docker containers or K8s). and a group or "container" can have associations to Docker, Podman K8s or things like that. In future may be web assembly containers. this could also help to generate out manifest to support multi cloud deployments or not?

**Comment by davidfowl** [comment]
Sounds great for the future, the only problem is that it would be in the far future 😅 as this is not the 90% case (AFAIK). Having given this more thought, by default when you add endpoints to containers thy go through our proxy. Even if the container is on a remote host the proxy is local.

cc @karolz-ms for ideas on short term and long term solutions

**Comment by biqas** [comment]
@davidfowl I have already tried those things:

appconfig:
`"AppHost": {
  "ContainerHostName": "bits.fedora"
}`

using this package: `<PackageReference Include="Aspire.Hosting" Version="8.0.0-preview.5.24161.15" />`

**Comment by karolz-ms** [comment]
@biqas can you tell me what is shown in the ports section of an inspected container when you run `podman container inspect` from your Windows VM on some container with mapped ports?

I am thinking this may be as simple as a bug in the Aspire orchestrator in the sense that it assumes containers will always be exposed via one of the loopback interfaces when configuring proxies.

**Comment by biqas** [comment]
Hi,

tried with version 9.1, still not working.

`Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler.Proxy[0]
      Error handling TCP connection     {"Service": {"name":"pg"}, "error": "Could not establish TCP connection to endpoint: dial tcp 127.0.0.1:40429: connectex: No connection could be made because the target machine actively refused it."}
fail: Aspire.Hosting.Dcp.dcpctrl.NetworkReconciler[0]
      could not connect a container to the network      {"NetworkName": {"name":"default-aspire-network"}, "Reconciliation": 24, "Container": "ee4280ad505bcd5535699a428a836f6485dfc9619a3a37f3a0a7a5ebe521d330", "Network": "default-aspire-network-0brnrm4c90", "error": "container ee4280ad505bcd5535699a428a836f6485dfc9619a3a37f3a0a7a5ebe521d330 is not connected to network 14610662f61e49aea845513754312f8c866f87766bade2fbced4ce4e8f76fbe8\nobject already exists\ncontainer already attached"}
fail: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler.Proxy[0]
      Error handling TCP connection     {"Service": {"name":"pg"}, "error": "Could not establish TCP connection to endpoint: dial tcp 127.0.0.1:40429: connectex: No connection could be made because the target machine actively refused it."}
fail: Aspire.Hosting.Dcp.dcpctrl.ServiceReconciler.Proxy[0]`

network "default-aspire-network-0brnrm4c..." is created and also the container is using the network.

`podman container inspect ee4280ad505b`:
`... 
"Networks": {
                    "default-aspire-network-0brnrm4c90": {
                         "EndpointID": "",
                         "Gateway": "10.89.0.1",
                         "IPAddress": "10.89.0.2",
                         "IPPrefixLen": 24,
                         "IPv6Gateway": "",
                         "GlobalIPv6Address": "",
                         "GlobalIPv6PrefixLen": 0,
                         "MacAddress": "22:43:03:36:0d:aa",
                         "NetworkID": "default-aspire-network-0brnrm4c90",
                         "DriverOpts": null,
                         "IPAMConfig": null,
                         "Links": null,
                         "Aliases": [
                              "pg",
                              "ee4280ad505b"
                         ]
                    }
               }
...`

**Comment by karolz-ms** [comment]
Right, we did not do any work to enable consumption of containers running on a separate machine (not bound to the loopback interface).

@biqas my question here https://github.com/dotnet/aspire/issues/2811#issuecomment-1995986769 was meant to tease out how you are exposing your containers running on the Linux VM to the Windows VM.

I also want to point out that what is supported today (on Mac) is running/debugging Aspire apps via Podman Desktop and VS Code.

**Comment by maddymontaquila** [comment]
Does this new feature in 9.4 help at all? https://learn.microsoft.com/en-us/dotnet/aspire/whats-new/dotnet-aspire-9.4#-enhanced-endpoint-url-support ( #10232  )

**Comment by davidfowl** [comment]
We need to clean up the API a bit, but it should work.

**Comment by maddymontaquila** [comment]
@biqas let me know if that works ^ otherwise this will auto close in a couple of weeks. Thanks!
